0000-settlement-engine/0000-settlement-engine.md

```diff
@@ -0,0 +1,353 @@
+Accept: application/json
+Content-Type: application/json
+```
+
+> **[`Quantity`](#quantity-type)** to settle
+>
+> - If the amount is `0`, the settlement engine MAY retry failed or queued settlements.
+
+##### Response
+
+```http
+HTTP/1.1 200 OK
+Content-Type: application/json
+```
+
+> **[`Quantity`](#quantity-type)** enqueued to settle
+>
+> - This response does not guarantee that a settlement was executed.
+> - If the settlement engine uses a unit less precise than the accounting system's unit, in order to prevent the two systems getting out-of-sync, it must indicate the amount it queued so the accounting system can correctly track the amount leftover.
+
+#### Handle incoming request
+
+Respond to an incoming message from the given peer's settlement engine. The transport MUST be the only entity invoking this instruction.
+
+##### Request
+
+```http
+POST /accounts/:id/handleMessage HTTP/1.1
+Accept: application/octet-stream
+Content-Type: application/octet-stream
+```
+
+> _&lt;raw bytes of message&gt;_
+
+##### Response
+
+```http
+HTTP/1.1 200 OK
+Content-Type: application/octet-stream
+```
+
+> _&lt;raw bytes of response&gt;_
+
+#### Complete specification
+
+<details>
+  <summary>The full settlement engine HTTP API, including all response codes, error codes, and additional endpoints for manual management are available in this Swagger definition.</summary>
+
+```
+TODO Embed the definition here
+```
+
+</details>
+
+Settlement engine implementations are RECOMMENDED to implement the standardized endpoints for manual management noted above.
+
+#### Additional RPCs
+
+Settlement engine implementations MAY expose additional, non-standard endpoints for manual operations or configuration.
+
+For example, settlement engines that use payment channels may expose functionality to manually open, fund, and close channels.
+
+### Accounting System HTTP API
+
+#### Credit incoming settlement
+
+Notify the accounting system of receipt of an incoming settlement. Note that the settlement engine MAY accrue incoming settlement acknowledgements without immediately informing the accounting system.
+
+The settlement engine MUST be the only entity invoking this message and MUST credit this amount to the accounts receivable, or some asset account tracking the balance owed by the peer, to ensure accurate accounting.
+
+##### Request
+
+```http
+POST /accounts/:id/receipt HTTP/1.1
+Accept: application/json
+Content-Type: application/json
+```
+
+> **[`Quantity`](#quantity-type)** that should be credited to the account as an incoming settlement
+
+##### Response
+
+```http
+HTTP/1.1 200 OK
+Content-Type: application/json
+```
+
+> **[`Quantity`](#quantity-type)** credited to the account
+>
+> - If the accounting system uses a unit less precise than the settlement engine's unit in order to prevent the two systems getting out-of-sync, it must indicate the amount it acknowledged receipt of so the settlement engine can track the amount leftover.
```

Comment:
All of this asymmetric precision stuff adds a lot of complication, is it necessary? If the accounting system uses a precision higher than the underlying ledger, the settlement engine could just use the higher precision and internally downgrade it to ledger precision when performing settlements.

Reply:
Agree. But settlement engine still needs to track the amount that's too precise to settle -- is the idea that the SE just uses some arbitrary precision field to persist that?

In the opposite case, if the SE used a more precise scale than the accounting system, the connector would need to store the leftover amount, which would require _it_ to also use an arbitrary precision field.

@emschwartz @sappenin @gakonst Thoughts on that approach to handle different scales (as opposed to including the scale in every request)?

Reply:
Using arbitrary precision numbers in the SE wouldn't be too expensive. But using them in the accounting system would require bignumber operations on every packet. If the SE could hit an endpoint on the accounting system to discover the scale the accounting system was working in, then it could just always use the accounting system's scale. Then the accounting system can use integers, which will be faster

Reply:
> But using them in the accounting system would require bignumber operations on every packet

Within the accounting system, the actual balance could still be fixed point. To handle incoming settlements, the rounded amount could be credited to the balance, and the leftover could be stored in an arbitrary precision field until more settlements are received.

Reply:
Resolving for now. To be addressed pending outcome of [this call](https://interledger.slack.com/archives/CHZC9MB88/p1564734224168900) and [this issue thread](https://github.com/emschwartz/interledger-rs/issues/185).